dotnet build run into 429 and ignore Retry-After Header

---
**Issue moved from dotnet/sdk#15619**
- Please respond to @fotto1.

---

_From @fotto1 on Tuesday, February 2, 2021 1:11:15 PM_

We using for our project docker image mcr.microsoft.com/dotnet/core/sdk:3.1 with digest mcr.microsoft.com/dotnet/core/sdk@sha256:0fece15a102530aa2dad9d247bc0d05db6790917696377fc56a8465604ef1aff to run dotnet cli.

Our nuget packages are stored on an gitlab package manager version 13.8

After there are to many http request nuget receives 429 following RFC6585 with Retry-After Header in Response like described in https://tools.ietf.org/html/rfc6585#section-4

Instead of do a retry later `dotnet build` continue with download next packages and exit with 1 after third download fails with 429

```logs
Microsoft (R) Build Engine version 16.7.2+b60ddb6f4 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.
  Determining projects to restore...
  Retrying 'FindPackagesByIdAsync' for source 'https://SERVER_URL/api/v4/projects/PROJECT_ID/packages/nuget/download/microsoft.extensions.configuration.binder/index.json'.
  Response status code does not indicate success: 429 (Too Many Requests).
  Retrying 'FindPackagesByIdAsync' for source 'https://SERVER_URL/api/v4/projects/PROJECT_ID/packages/nuget/download/microsoft.extensions.configuration.binder/index.json'.
  Response status code does not indicate success: 429 (Too Many Requests).
  Retrying 'FindPackagesByIdAsync' for source 'https://SERVER_URL/api/v4/projects/PROJECT_ID/packages/nuget/download/microsoft.extensions.configuration.environmentvariables/index.json'.
  Response status code does not indicate success: 429 (Too Many Requests).
```

Can you please check how 429 is handled in `dotnet build` in case nuget is configured

Memorializing the failure in question here, anyway, until we have data on whether it's common --
```
Process ID: 3712
    Microsoft.NET.Build.Tests.GivenThatWeWantToUseAnalyzers.It_resolves_multitargeted_analyzers [FAIL]
      Expected command to pass but it did not.
      File Name: C:\h\w\BDDE09E9\p\d\dotnet.exe
      Arguments: msbuild /t:WriteValuesToFile C:\h\w\BDDE09E9\t\dotnetSdkTests\1v3w1tea.knb\It_resolves_m---47570CA8\It_resolves_multitargeted_analyzers\It_resolves_multitargeted_analyzers.csproj /p:ValueName=Analyzer /restore -p:TargetFramework=net472
      Exit Code: 1
      StdOut:
Running C:\h\w\BDDE09E9\p\d\dotnet.exe msbuild /t:Build C:\h\w\BDDE09E9\t\dotnetSdkTests\1v3w1tea.knb\It_handles_co---209447AF\ContentFiles\ContentFiles.csproj /restore
      MSBuild version 17.4.0-preview-22458-02+a3b647d76 for .NET
        Determining projects to restore...
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v3/flat2/system.runtime.compilerservices.unsafe/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/49a1bb2b-12b0-475f-adbd-1560fc76be38/nuget/v3/flat2/microsoft.bcl.asyncinterfaces/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/516521bf-6417-457e-9a9c-0a4bdfde03e7/nuget/v3/flat2/microsoft.bcl.asyncinterfaces/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/4828dfac-e9f8-49bc-acb6-319be99331fc/nuget/v3/flat2/system.buffers/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/system.memory/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/4828dfac-e9f8-49bc-acb6-319be99331fc/nuget/v3/flat2/microsoft.bcl.asyncinterfaces/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.threading.tasks.extensions/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/system.buffers/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/flat2/microsoft.bcl.asyncinterfaces/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/516521bf-6417-457e-9a9c-0a4bdfde03e7/nuget/v3/flat2/system.runtime.compilerservices.unsafe/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/516521bf-6417-457e-9a9c-0a4bdfde03e7/nuget/v3/flat2/system.text.encodings.web/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/system.runtime.compilerservices.unsafe/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/49e5305d-d845-4a14-9d69-6f5dbfb9570c/nuget/v3/flat2/microsoft.bcl.asyncinterfaces/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.valuetuple/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v3/flat2/system.memory/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/e31c6eea-0277-49f3-8194-142be67a9f72/nuget/v3/flat2/system.valuetuple/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/flat2/system.numerics.vectors/index.json'.
        Response status code does not indicate success: 429 ().
        Failed to download package 'System.Numerics.Vectors.4.5.0' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.numerics.vectors/4.5.0/system.numerics.vectors.4.5.0.nupkg'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.numerics.vectors/index.json'.
        Response status code does not indicate success: 429 ().
        Failed to download package 'System.Numerics.Vectors.4.5.0' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.numerics.vectors/4.5.0/system.numerics.vectors.4.5.0.nupkg'.
        Response status code does not indicate success: 429 ().
        Failed to download package 'System.Memory.4.5.3' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.memory/4.5.3/system.memory.4.5.3.nupkg'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/system.buffers/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.numerics.vectors/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v3/flat2/system.memory/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/flat2/system.numerics.vectors/index.json'.
        Response status code does not indicate success: 429 ().
        Failed to download package 'System.Memory.4.5.3' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.memory/4.5.3/system.memory.4.5.3.nupkg'.
        Response status code does not indicate success: 429 ().
        Failed to download package 'System.Numerics.Vectors.4.5.0' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.numerics.vectors/4.5.0/system.numerics.vectors.4.5.0.nupkg'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.numerics.vectors/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/flat2/system.numerics.vectors/index.json'.
Process ID: 6920
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v3/flat2/system.memory/index.json'.
        Response status code does not indicate success: 429 ().
        Failed to download package 'System.Memory.4.5.3' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.memory/4.5.3/system.memory.4.5.3.nupkg'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/system.buffers/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.numerics.vectors/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/1a5f89f6-d8da-4080-b15f-242650c914a8/nuget/v3/flat2/system.numerics.vectors/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v3/flat2/system.memory/index.json'.
        Response status code does not indicate success: 429 ().
        Failed to download package 'System.Memory.4.5.3' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.memory/4.5.3/system.memory.4.5.3.nupkg'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.numerics.vectors/index.json'.
        Response status code does not indicate success: 429 ().
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/system.buffers/index.json'.
        Response status code does not indicate success: 429 ().
        Failed to download package 'System.Memory.4.5.3' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/system.memory/4.5.3/system.memory.4.5.3.nupkg'.
        Response status code does not indicate success: 429 ().
      C:\h\w\BDDE09E9\t\dotnetSdkTests\1v3w1tea.knb\It_resolves_m---47570CA8\It_resolves_multitargeted_analyzers\It_resolves_multitargeted_analyzers.csproj : error NU1301: Failed to retrieve information about 'System.Numerics.Vectors' from remote source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.numerics.vectors/index.json'. [TargetFramework=net472]
      C:\h\w\BDDE09E9\t\dotnetSdkTests\1v3w1tea.knb\It_resolves_m---47570CA8\It_resolves_multitargeted_analyzers\It_resolves_multitargeted_analyzers.csproj : error NU1301: Failed to retrieve information about 'System.Numerics.Vectors' from remote source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/c9f8ac11-6bd8-4926-8306-f075241547f7/nuget/v3/flat2/system.numerics.vectors/index.json'. [TargetFramework=net472]
        Retrying 'FindPackagesByIdAsync' for source 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/system.buffers/index.json'.
        Response status code does not indicate success: 429 ().
        Failed to restore C:\h\w\BDDE09E9\t\dotnetSdkTests\1v3w1tea.knb\It_resolves_m---47570CA8\It_resolves_multitargeted_analyzers\It_resolves_multitargeted_analyzers.csproj (in 1.67 min).
```
For completeness we do not know of course whether doing anything special for 429 for this server would help. Rerunning the build succeeded.

--- a/src/NuGet.Core/NuGet.Protocol/EnhancedHttpRetryHelper.cs
+++ b/src/NuGet.Core/NuGet.Protocol/EnhancedHttpRetryHelper.cs
@@ -63,6 +63,11 @@ namespace NuGet.Protocol
         /// </summary>
         public const string ObserveRetryAfterEnvironmentVariableName = "NUGET_OBSERVE_RETRY_AFTER";
 
+        /// <summary>
+        /// The environment variable used to set maximum Retry-After delay period
+        /// </summary>
+        public const string MaximumRetryAfterDurationEnvironmentVariableName = "NUGET_MAX_RETRY_AFTER_DELAY_SECONDS";
+
         private readonly IEnvironmentVariableReader _environmentVariableReader;
 
         private bool? _isEnabled = null;
@@ -74,6 +79,9 @@ namespace NuGet.Protocol
         private bool? _retry429 = null;
 
         private bool? _observeRetryAfter = null;
+
+        private TimeSpan? _maxRetyAfterDelay = null;
+        private bool _gotMaxRetryAfterDelay = false;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="EnhancedHttpRetryHelper" /> class.
@@ -109,6 +117,24 @@ namespace NuGet.Protocol
         /// Gets a value indicating whether or not to observe the Retry-After header on HTTP responses.
         /// </summary>
         internal bool ObserveRetryAfter => _observeRetryAfter ??= GetBoolFromEnvironmentVariable(ObserveRetryAfterEnvironmentVariableName, defaultValue: DefaultObserveRetryAfter, _environmentVariableReader);
+
+        internal TimeSpan? MaxRetryAfterDelay
+        {
+            get
+            {
+                if (!_gotMaxRetryAfterDelay)
+                {
+                    if (int.TryParse(_environmentVariableReader.GetEnvironmentVariable(MaximumRetryAfterDurationEnvironmentVariableName), out int maxRetryAfterDelay))
+                    {
+                        _maxRetyAfterDelay = TimeSpan.FromSeconds(maxRetryAfterDelay);
+                    }
+
+                    _gotMaxRetryAfterDelay = true;
+                }
+
+                return _maxRetyAfterDelay;
+            }
+        }
 
         /// <summary>
         /// Gets a <see cref="bool" /> value from the specified environment variable.

--- a/src/NuGet.Core/NuGet.Protocol/EnhancedHttpRetryHelper.cs
+++ b/src/NuGet.Core/NuGet.Protocol/EnhancedHttpRetryHelper.cs
@@ -34,6 +34,11 @@ namespace NuGet.Protocol
         public const bool DefaultRetry429 = true;
 
         /// <summary>
+        /// The default value indicating whether or not to observe Retry-After headers on responses.
+        /// </summary>
+        public const bool DefaultObserveRetryAfter = true;
+
+        /// <summary>
         /// The environment variable used to change the delay value.
         /// </summary>
         public const string DelayInMillisecondsEnvironmentVariableName = "NUGET_ENHANCED_NETWORK_RETRY_DELAY_MILLISECONDS";
@@ -53,6 +58,11 @@ namespace NuGet.Protocol
         /// </summary>
         public const string Retry429EnvironmentVariableName = "NUGET_RETRY_HTTP_429";
 
+        /// <summary>
+        /// The envionment variable to disable observing Retry-After responses.
+        /// </summary>
+        public const string ObserveRetryAfterEnvironmentVariableName = "NUGET_OBSERVE_RETRY_AFTER";
+
         private readonly IEnvironmentVariableReader _environmentVariableReader;
 
         private bool? _isEnabled = null;
@@ -62,6 +72,8 @@ namespace NuGet.Protocol
         private int? _delayInMilliseconds = null;
 
         private bool? _retry429 = null;
+
+        private bool? _observeRetryAfter = null;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="EnhancedHttpRetryHelper" /> class.
@@ -92,6 +104,11 @@ namespace NuGet.Protocol
         /// Gets a value indicating whether or not retryable HTTP 4xx responses should be retried.
         /// </summary>
         internal bool Retry429 => _retry429 ??= GetBoolFromEnvironmentVariable(Retry429EnvironmentVariableName, defaultValue: true, _environmentVariableReader);
+
+        /// <summary>
+        /// Gets a value indicating whether or not to observe the Retry-After header on HTTP responses.
+        /// </summary>
+        internal bool ObserveRetryAfter => _observeRetryAfter ??= GetBoolFromEnvironmentVariable(ObserveRetryAfterEnvironmentVariableName, defaultValue: DefaultObserveRetryAfter, _environmentVariableReader);
 
         /// <summary>
         /// Gets a <see cref="bool" /> value from the specified environment variable.

--- a/src/NuGet.Core/NuGet.Protocol/EnhancedHttpRetryHelper.cs
+++ b/src/NuGet.Core/NuGet.Protocol/EnhancedHttpRetryHelper.cs
@@ -81,7 +81,6 @@ namespace NuGet.Protocol
         private bool? _observeRetryAfter = null;
 
         private TimeSpan? _maxRetyAfterDelay = null;
-        private bool _gotMaxRetryAfterDelay = false;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="EnhancedHttpRetryHelper" /> class.
@@ -118,21 +117,17 @@ namespace NuGet.Protocol
         /// </summary>
         internal bool ObserveRetryAfter => _observeRetryAfter ??= GetBoolFromEnvironmentVariable(ObserveRetryAfterEnvironmentVariableName, defaultValue: DefaultObserveRetryAfter, _environmentVariableReader);
 
-        internal TimeSpan? MaxRetryAfterDelay
+        internal TimeSpan MaxRetryAfterDelay
         {
             get
             {
-                if (!_gotMaxRetryAfterDelay)
+                if (_maxRetyAfterDelay == null)
                 {
-                    if (int.TryParse(_environmentVariableReader.GetEnvironmentVariable(MaximumRetryAfterDurationEnvironmentVariableName), out int maxRetryAfterDelay))
-                    {
-                        _maxRetyAfterDelay = TimeSpan.FromSeconds(maxRetryAfterDelay);
-                    }
-
-                    _gotMaxRetryAfterDelay = true;
+                    int maxRetryAfterDelay = GetIntFromEnvironmentVariable(MaximumRetryAfterDurationEnvironmentVariableName, defaultValue: (int)TimeSpan.FromHours(1).TotalSeconds, _environmentVariableReader);
+                    _maxRetyAfterDelay = TimeSpan.FromSeconds(maxRetryAfterDelay);
                 }
 
-                return _maxRetyAfterDelay;
+                return _maxRetyAfterDelay.Value;
             }
         }
 

--- a/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs
+++ b/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs
@@ -40,12 +40,10 @@ namespace NuGet.Protocol
         /// requests cannot always be used. For example, suppose the request is a POST and contains content
         /// of a stream that can only be consumed once.
         /// </remarks>
-#nullable disable
         public Task<HttpResponseMessage> SendAsync(
             HttpRetryHandlerRequest request,
             ILogger log,
             CancellationToken cancellationToken)
-#nullable enable
         {
             return SendAsync(request, source: string.Empty, log, cancellationToken);
         }
@@ -58,13 +56,11 @@ namespace NuGet.Protocol
         /// requests cannot always be used. For example, suppose the request is a POST and contains content
         /// of a stream that can only be consumed once.
         /// </remarks>
-#nullable disable
         public async Task<HttpResponseMessage> SendAsync(
             HttpRetryHandlerRequest request,
             string source,
             ILogger log,
             CancellationToken cancellationToken)
-#nullable enable
         {
             if (source == null)
             {
@@ -79,10 +75,9 @@ namespace NuGet.Protocol
 
             var tries = 0;
             HttpResponseMessage? response = null;
-            var success = false;
             TimeSpan? retryAfter = null;
 
-            while (tries < request.MaxTries && !success)
+            while (true)
             {
                 // There are many places where another variable named "MaxTries" is set to 1,
                 // so the Delay() never actually occurs.
@@ -112,7 +107,6 @@ namespace NuGet.Protocol
                 }
 
                 tries++;
-                success = true;
 
                 using (var requestMessage = request.RequestFactory())
                 {
@@ -241,7 +235,14 @@ namespace NuGet.Protocol
                         // 429 == too many requests
                         if (statusCode >= 500 || ((statusCode == 408 || statusCode == 429) && _enhancedHttpRetryHelper.Retry429))
                         {
-                            success = false;
+                            if (tries == request.MaxTries)
+                            {
+                                return response;
+                            }
+                        }
+                        else
+                        {
+                            return response;
                         }
                     }
                     catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
@@ -265,8 +266,6 @@ namespace NuGet.Protocol
                     }
                     catch (Exception e)
                     {
-                        success = false;
-
                         response?.Dispose();
 
                         ProtocolDiagnostics.RaiseEvent(new ProtocolDiagnosticHttpEvent(
@@ -298,8 +297,6 @@ namespace NuGet.Protocol
                     }
                 }
             }
-
-            return response;
         }
 
         private static TimeSpan? GetRetryAfter(RetryConditionHeaderValue? retryAfter)

--- a/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs
+++ b/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs
@@ -213,6 +213,15 @@ namespace NuGet.Protocol
                                 requestUri,
                                 bodyStopwatch.ElapsedMilliseconds,
                                 retryAfter.Value.TotalSeconds));
+
+                            if (retryAfter.Value.TotalMilliseconds < 0)
+                            {
+                                retryAfter = null;
+                            }
+                            else if (retryAfter.Value > _enhancedHttpRetryHelper.MaxRetryAfterDelay)
+                            {
+                                retryAfter = _enhancedHttpRetryHelper.MaxRetryAfterDelay;
+                            }
                         }
                         else
                         {
@@ -222,11 +231,6 @@ namespace NuGet.Protocol
                                 response.StatusCode,
                                 requestUri,
                                 bodyStopwatch.ElapsedMilliseconds));
-                        }
-
-                        if (retryAfter?.TotalMilliseconds < 0)
-                        {
-                            retryAfter = null;
                         }
 
                         int statusCode = (int)response.StatusCode;

--- a/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs
+++ b/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs
@@ -1,11 +1,14 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+#nullable enable
+
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Globalization;
 using System.Net.Http;
+using System.Net.Http.Headers;
 using System.Threading;
 using System.Threading.Tasks;
 using NuGet.Common;
@@ -37,10 +40,12 @@ namespace NuGet.Protocol
         /// requests cannot always be used. For example, suppose the request is a POST and contains content
         /// of a stream that can only be consumed once.
         /// </remarks>
+#nullable disable
         public Task<HttpResponseMessage> SendAsync(
             HttpRetryHandlerRequest request,
             ILogger log,
             CancellationToken cancellationToken)
+#nullable enable
         {
             return SendAsync(request, source: string.Empty, log, cancellationToken);
         }
@@ -53,11 +58,13 @@ namespace NuGet.Protocol
         /// requests cannot always be used. For example, suppose the request is a POST and contains content
         /// of a stream that can only be consumed once.
         /// </remarks>
+#nullable disable
         public async Task<HttpResponseMessage> SendAsync(
             HttpRetryHandlerRequest request,
             string source,
             ILogger log,
             CancellationToken cancellationToken)
+#nullable enable
         {
             if (source == null)
             {
@@ -71,8 +78,9 @@ namespace NuGet.Protocol
             }
 
             var tries = 0;
-            HttpResponseMessage response = null;
+            HttpResponseMessage? response = null;
             var success = false;
+            TimeSpan? retryAfter = null;
 
             while (tries < request.MaxTries && !success)
             {
@@ -80,7 +88,11 @@ namespace NuGet.Protocol
                 // so the Delay() never actually occurs.
                 // When opted in to "enhanced retry", do the delay and have it increase exponentially where applicable
                 // (i.e. when "tries" is allowed to be > 1)
-                if (tries > 0 || (_enhancedHttpRetryHelper.IsEnabled && request.IsRetry))
+                if (retryAfter != null && _enhancedHttpRetryHelper.ObserveRetryAfter)
+                {
+                    await Task.Delay(retryAfter.Value, cancellationToken);
+                }
+                else if (tries > 0 || (_enhancedHttpRetryHelper.IsEnabled && request.IsRetry))
                 {
                     // "Enhanced" retry: In the case where this is actually a 2nd-Nth try, back off exponentially with some random.
                     // In many cases due to the external retry loop, this will be always be 1 * request.RetryDelay.TotalMilliseconds + 0-200 ms
@@ -107,7 +119,7 @@ namespace NuGet.Protocol
                     var stopwatches = new List<Stopwatch>(2);
                     var bodyStopwatch = new Stopwatch();
                     stopwatches.Add(bodyStopwatch);
-                    Stopwatch headerStopwatch = null;
+                    Stopwatch? headerStopwatch = null;
                     if (request.CompletionOption == HttpCompletionOption.ResponseHeadersRead)
                     {
                         headerStopwatch = new Stopwatch();
@@ -118,7 +130,7 @@ namespace NuGet.Protocol
 #else
                     requestMessage.Properties[StopwatchPropertyName] = stopwatches;
 #endif
-                    var requestUri = requestMessage.RequestUri;
+                    var requestUri = requestMessage.RequestUri!;
 
                     try
                     {
@@ -197,12 +209,31 @@ namespace NuGet.Protocol
                             response.Content = newContent;
                         }
 
-                        log.LogInformation("  " + string.Format(
-                            CultureInfo.InvariantCulture,
-                            Strings.Http_ResponseLog,
-                            response.StatusCode,
-                            requestUri,
-                            bodyStopwatch.ElapsedMilliseconds));
+                        retryAfter = GetRetryAfter(response.Headers.RetryAfter);
+                        if (retryAfter != null)
+                        {
+                            log.LogInformation("  " + string.Format(
+                                CultureInfo.InvariantCulture,
+                                Strings.Http_ResponseLogWithRetryAfter,
+                                response.StatusCode,
+                                requestUri,
+                                bodyStopwatch.ElapsedMilliseconds,
+                                retryAfter.Value.TotalSeconds));
+                        }
+                        else
+                        {
+                            log.LogInformation("  " + string.Format(
+                                CultureInfo.InvariantCulture,
+                                Strings.Http_ResponseLog,
+                                response.StatusCode,
+                                requestUri,
+                                bodyStopwatch.ElapsedMilliseconds));
+                        }
+
+                        if (retryAfter?.TotalMilliseconds < 0)
+                        {
+                            retryAfter = null;
+                        }
 
                         int statusCode = (int)response.StatusCode;
                         // 5xx == server side failure
@@ -269,6 +300,23 @@ namespace NuGet.Protocol
             }
 
             return response;
+        }
+
+        private static TimeSpan? GetRetryAfter(RetryConditionHeaderValue? retryAfter)
+        {
+            if (retryAfter?.Delta != null)
+            {
+                return retryAfter.Delta;
+            }
+
+            if (retryAfter?.Date != null)
+            {
+                DateTimeOffset retryAfterDate = retryAfter.Date.Value.ToUniversalTime();
+                var now = DateTimeOffset.UtcNow;
+                return retryAfterDate - now;
+            }
+
+            return null;
         }
     }
 }

--- a/src/NuGet.Core/NuGet.Protocol/PublicAPI/net472/PublicAPI.Shipped.txt
+++ b/src/NuGet.Core/NuGet.Protocol/PublicAPI/net472/PublicAPI.Shipped.txt
@@ -459,8 +459,8 @@ NuGet.Protocol.HttpRequestMessageFactory
 NuGet.Protocol.HttpResponseMessageExtensions
 NuGet.Protocol.HttpRetryHandler
 NuGet.Protocol.HttpRetryHandler.HttpRetryHandler() -> void
-~NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest request, NuGet.Common.ILogger log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage>
-~NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest request, string source, NuGet.Common.ILogger log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage>
+NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest! request, NuGet.Common.ILogger! log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage!>!
+NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest! request, string! source, NuGet.Common.ILogger! log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage!>!
 NuGet.Protocol.HttpRetryHandlerRequest
 ~NuGet.Protocol.HttpRetryHandlerRequest.AddHeaders.get -> System.Collections.Generic.IList<System.Collections.Generic.KeyValuePair<string, System.Collections.Generic.IEnumerable<string>>>
 ~NuGet.Protocol.HttpRetryHandlerRequest.AddHeaders.set -> void

--- a/src/NuGet.Core/NuGet.Protocol/PublicAPI/netcoreapp5.0/PublicAPI.Shipped.txt
+++ b/src/NuGet.Core/NuGet.Protocol/PublicAPI/netcoreapp5.0/PublicAPI.Shipped.txt
@@ -459,8 +459,8 @@ NuGet.Protocol.HttpRequestMessageFactory
 NuGet.Protocol.HttpResponseMessageExtensions
 NuGet.Protocol.HttpRetryHandler
 NuGet.Protocol.HttpRetryHandler.HttpRetryHandler() -> void
-~NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest request, NuGet.Common.ILogger log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage>
-~NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest request, string source, NuGet.Common.ILogger log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage>
+NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest! request, NuGet.Common.ILogger! log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage!>!
+NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest! request, string! source, NuGet.Common.ILogger! log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage!>!
 NuGet.Protocol.HttpRetryHandlerRequest
 ~NuGet.Protocol.HttpRetryHandlerRequest.AddHeaders.get -> System.Collections.Generic.IList<System.Collections.Generic.KeyValuePair<string, System.Collections.Generic.IEnumerable<string>>>
 ~NuGet.Protocol.HttpRetryHandlerRequest.AddHeaders.set -> void

--- a/src/NuGet.Core/NuGet.Protocol/PublicAPI/netstandard2.0/PublicAPI.Shipped.txt
+++ b/src/NuGet.Core/NuGet.Protocol/PublicAPI/netstandard2.0/PublicAPI.Shipped.txt
@@ -459,8 +459,8 @@ NuGet.Protocol.HttpRequestMessageFactory
 NuGet.Protocol.HttpResponseMessageExtensions
 NuGet.Protocol.HttpRetryHandler
 NuGet.Protocol.HttpRetryHandler.HttpRetryHandler() -> void
-~NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest request, NuGet.Common.ILogger log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage>
-~NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest request, string source, NuGet.Common.ILogger log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage>
+NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest! request, NuGet.Common.ILogger! log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage!>!
+NuGet.Protocol.HttpRetryHandler.SendAsync(NuGet.Protocol.HttpRetryHandlerRequest! request, string! source, NuGet.Common.ILogger! log, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage!>!
 NuGet.Protocol.HttpRetryHandlerRequest
 ~NuGet.Protocol.HttpRetryHandlerRequest.AddHeaders.get -> System.Collections.Generic.IList<System.Collections.Generic.KeyValuePair<string, System.Collections.Generic.IEnumerable<string>>>
 ~NuGet.Protocol.HttpRetryHandlerRequest.AddHeaders.set -> void

--- a/src/NuGet.Core/NuGet.Protocol/Strings.Designer.cs
+++ b/src/NuGet.Core/NuGet.Protocol/Strings.Designer.cs
@@ -214,6 +214,15 @@ namespace NuGet.Protocol {
         }
         
         /// <summary>
+        ///   Looks up a localized string similar to {0} {1} {2}ms (retry after: {3}s).
+        /// </summary>
+        internal static string Http_ResponseLogWithRetryAfter {
+            get {
+                return ResourceManager.GetString("Http_ResponseLogWithRetryAfter", resourceCulture);
+            }
+        }
+        
+        /// <summary>
         ///   Looks up a localized string similar to The HTTP request to &apos;{0} {1}&apos; has timed out after {2}ms..
         /// </summary>
         internal static string Http_Timeout {
@@ -961,7 +970,7 @@ namespace NuGet.Protocol {
         }
         
         /// <summary>
-        ///   Looks up a localized string similar to You are running the &apos;{0}&apos; operation with an &apos;HTTP&apos; source, &apos;{1}&apos;. Non-HTTPS access will be removed in a future version. Consider migrating to &apos;HTTPS&apos; sources..
+        ///   Looks up a localized string similar to You are running the &apos;{0}&apos; operation with an &apos;HTTP&apos; source, &apos;{1}&apos;. Non-HTTPS access will be removed in a future version. Consider migrating to an &apos;HTTPS&apos; source..
         /// </summary>
         internal static string Warning_HttpServerUsage {
             get {

--- a/src/NuGet.Core/NuGet.Protocol/Strings.resx
+++ b/src/NuGet.Core/NuGet.Protocol/Strings.resx
@@ -482,4 +482,10 @@ The "ms" should be localized to the abbreviation for milliseconds.</comment>
     <value>You are running the '{0}' operation with an 'HTTP' source, '{1}'. Non-HTTPS access will be removed in a future version. Consider migrating to an 'HTTPS' source.</value>
     <comment>0 - The command name. Ex. Push/Delete. 1 - server uri.</comment>
   </data>
+  <data name="Http_ResponseLogWithRetryAfter" xml:space="preserve">
+    <value>{0} {1} {2}ms (retry after: {3}s)</value>
+    <comment>Parameters in order: non-localizable HTTP status, non-localizable HTTP request URI, request duration in milliseconds, retry after duration in milliseconds
+The "ms" should be localized to the abbreviation for milliseconds.
+The "s" should be localized to the abbreviation for seconds.</comment>
+  </data>
 </root>

--- a/test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/HttpRetryHandlerTests.cs
+++ b/test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/HttpRetryHandlerTests.cs
@@ -179,7 +179,7 @@ namespace NuGet.Core.FuncTest
         }
 
         [Fact]
-        public async Task HttpRetryHandler_429WithRetryAfterHeader_UsesHeaderDelay()
+        public async Task HttpRetryHandler_429WithRetryAfterHeader_UsesRetryAfterValue()
         {
             // Arrange
             int attempt = 0;
@@ -188,16 +188,10 @@ namespace NuGet.Core.FuncTest
 
             var busyServer = new HttpRetryTestHandler((req, cancellationToken) =>
             {
-                if (attempt++ < 2)
-                {
-                    var response = new HttpResponseMessage(tooManyRequestsStatusCode);
-                    response.Headers.RetryAfter = new System.Net.Http.Headers.RetryConditionHeaderValue(TimeSpan.FromMilliseconds(1));
-                    return Task.FromResult(response);
-                }
-                else
-                {
-                    return Task.FromResult(new HttpResponseMessage(HttpStatusCode.OK));
-                }
+                attempt++;
+                var response = new HttpResponseMessage(tooManyRequestsStatusCode);
+                response.Headers.RetryAfter = new System.Net.Http.Headers.RetryConditionHeaderValue(TimeSpan.FromMilliseconds(1));
+                return Task.FromResult(response);
             });
             var client = new HttpClient(busyServer);
 
@@ -227,7 +221,7 @@ namespace NuGet.Core.FuncTest
             }
 
             // Assert
-            response.StatusCode.Should().Be(HttpStatusCode.OK);
+            response.StatusCode.Should().Be(tooManyRequestsStatusCode);
             attempt.Should().Be(3);
         }
 

--- a/test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/HttpRetryHandlerTests.cs
+++ b/test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/HttpRetryHandlerTests.cs
@@ -3,10 +3,12 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Net;
 using System.Net.Http;
 using System.Threading;
 using System.Threading.Tasks;
+using FluentAssertions;
 using NuGet.Common;
 using NuGet.Protocol;
 using NuGet.Test.Server;
@@ -18,7 +20,7 @@ namespace NuGet.Core.FuncTest
 {
     public class HttpRetryHandlerTests
     {
-        private const string TestUrl = "https://test.local/test.json";
+        private const string TestUrl = "https://local.test/test.json";
 
         [Fact]
         public async Task HttpRetryHandler_ReturnsContentHeaders()
@@ -176,6 +178,59 @@ namespace NuGet.Core.FuncTest
 #endif
         }
 
+        [Fact]
+        public async Task HttpRetryHandler_429WithRetryAfterHeader_UsesHeaderDelay()
+        {
+            // Arrange
+            int attempt = 0;
+            // .NET Framework don't have HttpStatusCode.TooManyRequests
+            HttpStatusCode tooManyRequestsStatusCode = (HttpStatusCode)429;
+
+            var busyServer = new HttpRetryTestHandler((req, cancellationToken) =>
+            {
+                if (attempt++ < 2)
+                {
+                    var response = new HttpResponseMessage(tooManyRequestsStatusCode);
+                    response.Headers.RetryAfter = new System.Net.Http.Headers.RetryConditionHeaderValue(TimeSpan.FromMilliseconds(1));
+                    return Task.FromResult(response);
+                }
+                else
+                {
+                    return Task.FromResult(new HttpResponseMessage(HttpStatusCode.OK));
+                }
+            });
+            var client = new HttpClient(busyServer);
+
+            // Use a very long timeout that will tell us that Retry-After wasn't being used.
+            TimeSpan retryDelay = TimeSpan.FromHours(5);
+            var httpRetryHandlerRequest = new HttpRetryHandlerRequest(client, () => new HttpRequestMessage(HttpMethod.Get, TestUrl))
+            {
+                RetryDelay = retryDelay
+            };
+            TestEnvironmentVariableReader environmentVariables = GetEnhancedHttpRetryEnvironmentVariables(retry429: true, observeRetryAfter: true, delayMilliseconds: (int)retryDelay.TotalMilliseconds);
+            var httpRetryHandler = new HttpRetryHandler(environmentVariables);
+            var logger = new TestLogger();
+
+            // Act
+            HttpResponseMessage response;
+            int cancellationDelay = Debugger.IsAttached ? int.MaxValue : 1000;
+            using (var cts = new CancellationTokenSource(millisecondsDelay: cancellationDelay))
+            {
+                try
+                {
+                    response = await httpRetryHandler.SendAsync(httpRetryHandlerRequest, logger, cts.Token);
+                }
+                catch (OperationCanceledException) when (cts.Token.IsCancellationRequested)
+                {
+                    throw new Exception("HTTP response Retry-After not used");
+                }
+            }
+
+            // Assert
+            response.StatusCode.Should().Be(HttpStatusCode.OK);
+            attempt.Should().Be(3);
+        }
+
         private static async Task<T> ThrowsException<T>(ITestServer server) where T : Exception
         {
             return await server.ExecuteAsync(async address =>
@@ -212,6 +267,21 @@ namespace NuGet.Core.FuncTest
                 return exception;
             });
         }
+
+        private static TestEnvironmentVariableReader GetEnhancedHttpRetryEnvironmentVariables(
+            bool? isEnabled = true,
+            int? retryCount = 5,
+            int? delayMilliseconds = 0,
+            bool? retry429 = true,
+            bool? observeRetryAfter = true)
+        => new TestEnvironmentVariableReader(
+            new Dictionary<string, string>()
+            {
+                [EnhancedHttpRetryHelper.IsEnabledEnvironmentVariableName] = isEnabled?.ToString(),
+                [EnhancedHttpRetryHelper.RetryCountEnvironmentVariableName] = retryCount?.ToString(),
+                [EnhancedHttpRetryHelper.DelayInMillisecondsEnvironmentVariableName] = delayMilliseconds?.ToString(),
+                [EnhancedHttpRetryHelper.Retry429EnvironmentVariableName] = retry429?.ToString(),
+            });
 
         private class CountingHandler : DelegatingHandler
         {

--- a/test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/HttpRetryHandlerTests.cs
+++ b/test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/HttpRetryHandlerTests.cs
@@ -213,7 +213,7 @@ namespace NuGet.Core.FuncTest
 
             // Act
             HttpResponseMessage response;
-            int cancellationDelay = Debugger.IsAttached ? int.MaxValue : 1000;
+            int cancellationDelay = Debugger.IsAttached ? int.MaxValue : 60 * 1000;
             using (var cts = new CancellationTokenSource(millisecondsDelay: cancellationDelay))
             {
                 try
@@ -229,6 +229,55 @@ namespace NuGet.Core.FuncTest
             // Assert
             response.StatusCode.Should().Be(HttpStatusCode.OK);
             attempt.Should().Be(3);
+        }
+
+        [Fact]
+        public async Task HttpRetryHandler_LongRetryAfterDuration_UsesEnvVarSetting()
+        {
+            // Arrange
+            int attempts = 0;
+            // .NET Framework don't have HttpStatusCode.TooManyRequests
+            HttpStatusCode tooManyRequestsStatusCode = (HttpStatusCode)429;
+
+            var busyServer = new HttpRetryTestHandler((req, cancellationToken) =>
+            {
+                attempts++;
+                var response = new HttpResponseMessage(tooManyRequestsStatusCode);
+                response.Headers.RetryAfter = new System.Net.Http.Headers.RetryConditionHeaderValue(TimeSpan.FromHours(1));
+                return Task.FromResult(response);
+            });
+            var client = new HttpClient(busyServer);
+
+            // Use a very long timeout that will tell us that Retry-After wasn't being used.
+            TimeSpan retryDelay = TimeSpan.FromHours(1);
+            var httpRetryHandlerRequest = new HttpRetryHandlerRequest(client, () => new HttpRequestMessage(HttpMethod.Get, TestUrl))
+            {
+                MaxTries = 2,
+                RetryDelay = retryDelay
+            };
+            TestEnvironmentVariableReader environmentVariables =
+                GetEnhancedHttpRetryEnvironmentVariables(retryCount: 2, retry429: true, observeRetryAfter: true, delayMilliseconds: (int)retryDelay.TotalMilliseconds, maxRetryAfterSeconds: 1);
+            var httpRetryHandler = new HttpRetryHandler(environmentVariables);
+            var logger = new TestLogger();
+
+            // Act
+            HttpResponseMessage response;
+            int cancellationDelay = Debugger.IsAttached ? int.MaxValue : 60 * 1000;
+            using (var cts = new CancellationTokenSource(millisecondsDelay: cancellationDelay))
+            {
+                try
+                {
+                    response = await httpRetryHandler.SendAsync(httpRetryHandlerRequest, logger, cts.Token);
+                }
+                catch (OperationCanceledException) when (cts.Token.IsCancellationRequested)
+                {
+                    throw new Exception("HTTP response Retry-After not used");
+                }
+            }
+
+            // Assert
+            response.StatusCode.Should().Be(tooManyRequestsStatusCode);
+            attempts.Should().Be(2);
         }
 
         private static async Task<T> ThrowsException<T>(ITestServer server) where T : Exception
@@ -273,7 +322,8 @@ namespace NuGet.Core.FuncTest
             int? retryCount = 5,
             int? delayMilliseconds = 0,
             bool? retry429 = true,
-            bool? observeRetryAfter = true)
+            bool? observeRetryAfter = true,
+            int? maxRetryAfterSeconds = null)
         => new TestEnvironmentVariableReader(
             new Dictionary<string, string>()
             {
@@ -281,6 +331,8 @@ namespace NuGet.Core.FuncTest
                 [EnhancedHttpRetryHelper.RetryCountEnvironmentVariableName] = retryCount?.ToString(),
                 [EnhancedHttpRetryHelper.DelayInMillisecondsEnvironmentVariableName] = delayMilliseconds?.ToString(),
                 [EnhancedHttpRetryHelper.Retry429EnvironmentVariableName] = retry429?.ToString(),
+                [EnhancedHttpRetryHelper.ObserveRetryAfterEnvironmentVariableName] = observeRetryAfter?.ToString(),
+                [EnhancedHttpRetryHelper.MaximumRetryAfterDurationEnvironmentVariableName] = maxRetryAfterSeconds?.ToString(),
             });
 
         private class CountingHandler : DelegatingHandler

